core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -93,33 +97,51 @@ private[kafka] object LogValidator extends Logging {
     batch
   }
 
-  private def validateBatch(firstBatch: RecordBatch, batch: RecordBatch, isFromClient: Boolean, toMagic: Byte): Unit = {
+  private def validateBatch(topicPartition: TopicPartition, firstBatch: RecordBatch, batch: RecordBatch, isFromClient: Boolean, toMagic: Byte, brokerTopicStats: BrokerTopicStats): Unit = {
     // batch magic byte should have the same magic as the first batch
-    if (firstBatch.magic() != batch.magic())
+    if (firstBatch.magic() != batch.magic()) {
+      brokerTopicStats.topicStats(topicPartition.topic).invalidMagicNumberRecordsPerSec.mark()
+      brokerTopicStats.allTopicsStats.invalidMagicNumberRecordsPerSec.mark()
       throw new InvalidRecordException(s"Batch magic ${batch.magic()} is not the same as the first batch'es magic byte ${firstBatch.magic()}")
+    }
 
     if (isFromClient) {
       if (batch.magic >= RecordBatch.MAGIC_VALUE_V2) {
         val countFromOffsets = batch.lastOffset - batch.baseOffset + 1
-        if (countFromOffsets <= 0)
+        if (countFromOffsets <= 0) {
+          brokerTopicStats.topicStats(topicPartition.topic).invalidOffsetOrSequenceRecordsPerSec.mark()
```

Comment:
Ditto as above: let's record at allTopicsStats level, and then add topic / partition into the InvalidRecordException error message.

Reply:
And ditto below.